hummingbot/strategy/spot_perpetual_arbitrage/spot_perpetual_arbitrage.py

```diff
@@ -378,46 +376,48 @@ def check_perpetual_budget_constraint(self, proposal: ArbProposal) -> bool:
 
         return True
 
-    def execute_arb_proposal(self, proposal: ArbProposal):
+    def execute_arb_proposal(self, proposal: ArbProposal, order_amount: Decimal):
         """
         Execute both sides of the arbitrage trades concurrently.
         :param proposal: the arbitrage proposal
         """
-        if proposal.order_amount == s_decimal_zero:
+        initial_perp_position = abs(self.perp_positions[0].amount) if self.perp_positions else None
+        if order_amount == s_decimal_zero:
             return
         spot_side = proposal.spot_side
         spot_order_fn = self.buy_with_specific_market if spot_side.is_buy else self.sell_with_specific_market
         side = "BUY" if spot_side.is_buy else "SELL"
         self.log_with_clock(
             logging.INFO,
-            f"Placing {side} order for {proposal.order_amount} {spot_side.market_info.base_asset} "
-            f"at {spot_side.market_info.market.display_name} at {spot_side.order_price} price"
+            f"Placing {side} order for {order_amount} {spot_side.market_info.base_asset} "
+            f"at {spot_side.market_info.market.display_name} at {spot_side.get_order_price(order_amount)} price"
         )
         spot_order_fn(
             spot_side.market_info,
-            proposal.order_amount,
+            order_amount,
             spot_side.market_info.market.get_taker_order_type(),
-            spot_side.order_price,
+            spot_side.get_order_price(order_amount),
         )
         perp_side = proposal.perp_side
         perp_order_fn = self.buy_with_specific_market if perp_side.is_buy else self.sell_with_specific_market
         side = "BUY" if perp_side.is_buy else "SELL"
         position_action = PositionAction.CLOSE if self._strategy_state == StrategyState.Opened else PositionAction.OPEN
         self.log_with_clock(
             logging.INFO,
-            f"Placing {side} order for {proposal.order_amount} {perp_side.market_info.base_asset} "
-            f"at {perp_side.market_info.market.display_name} at {perp_side.order_price} price to "
+            f"Placing {side} order for {order_amount} {perp_side.market_info.base_asset} "
+            f"at {perp_side.market_info.market.display_name} at {perp_side.get_order_price(order_amount)} price to "
             f"{position_action.name} position."
         )
         perp_order_fn(
             perp_side.market_info,
-            proposal.order_amount,
+            order_amount,
             perp_side.market_info.market.get_taker_order_type(),
-            perp_side.order_price,
+            perp_side.get_order_price(order_amount),
             position_action=position_action
         )
         if self._strategy_state == StrategyState.Opened:
-            self._strategy_state = StrategyState.Closing
+            # possible partial closure of position
+            self._strategy_state = StrategyState.Closing if initial_perp_position == order_amount else StrategyState.Opening
```

Comment:
If its `Opened` but we execute an order amount smaller than current position, we need to keep position `Opened` afterwards.